File: 프로그래머스/lv3/43105. 정수 삼각형/정수 삼각형.py
N = -1
dp = []
def solve(x, y, triangle):
    if x >= N or x < y:
        return 0
    
    if dp[x][y] != -1:
        return dp[x][y]
    
    dp[x][y] = max(solve(x + 1, y, triangle), solve(x + 1, y + 1, triangle)) + triangle[x][y]
    
    return dp[x][y]
    
def solution(triangle):
    global dp, N
    N = len(triangle)
    dp = [[-1 for _ in range(i + 1)] for i in range(len(triangle))]
    return solve(0, 0, triangle)

File: 프로그래머스/3/43105. 정수 삼각형/정수 삼각형.java
import java.util.*;

class Solution {
    public int[][] triangle;
    public int[][] cache;
    
    public int solution(int[][] tg) {
        triangle = tg;
        
        cache = new int[tg.length][tg.length];
        for(int i = 0; i < tg.length; i++){
            Arrays.fill(cache[i], -1);
        }
        return solve(0,0);
    }
    
    public int solve(int i, int j){
        if(cache[i][j] != -1){
            return cache[i][j];
        }
        if(i == cache.length - 1){
            return triangle[i][j];
        }
        int result = triangle[i][j] + Math.max(solve(i + 1, j), solve(i + 1, j + 1));
        
        cache[i][j] = result;
        return result;
    }
}